README.md
# 《标题日记》 后台 - nodejs 版


## 一、项目说明
服务对象： [《标题日记》](https://github.com/KyleBing/diary-vue) 

> 线上已运行的例子：

> http [http://kylebing.cn:3000/diary/detail?diaryId=6766](http://kylebing.cn:3000/diary/detail?diaryId=6766)



### 
```bash
# 1. 统计信息
- /statistics # 统计信息 `2022-05-12`

# 2. 日记管理
- /diary/list
- /diary/add
- /diary/modify
- /diary/delete
- /diary/detail

    # 2.1 银行卡列表信息
    - /bank-card/ # 银行卡列表 `2022-05-12`
    
    # 2.2 账单统计信息
    - /bill/ # 日记中的账单类别统计信息 `2022-05-24`


# 3. 用户管理
- /user/
- /user/add
- /user/delete
- /user/modify
- /user/detail
- /user/list
- /user/login
- /user/change-password

# 4. 日记类别管理
- /diary-category/ # 类别管理 `2022-05-24`
- /diary-category/add # 类别管理 `2022-05-24`
- /diary-category/delete
- /diary-category/modify
- /diary-category/list
```



## 二、安装说明

**服务器需要的条件：**
- 已安装 `nodejs 16+`
- 已安装 `npm` 或 `yarn`
- 已安装 `nginx`，需要用它进行路径映射，以供前端非跨域式访问后台
- 已安装 `mysql` 或 `mariaDB`

### 1. clone 或 下载该项目文件
下载到项目文件后，执行 `npm i` 或者 `yarn` 安装项目依赖

### 2. 修改数据库配置文件
修改 `/config/configDatabase.js` 文件内容，改成你的配置
```js
module.exports = {
    host:       'localhost',
    user:       'root',
    password:   '----',
    port:       '3306',
    database:   'diary',
    multipleStatements: true
}
```

和项目配置文件 `/config/configProject.js`
```js
module.exports = {
    invitation: '----', // 邀请码，用于注册时使用
    adminCount: 'kylebing@163.com', // 管理员账户，该用户可以在统计页面中查看所有用户统计数据
    TOKEN_NAME: 'Diary-Token', // header 中的 token 字段
}
```

### 3. 启动程序
这里推荐使用 pm2 管理程序，创建名为 `diary` 的 pm2 项目，并启动
> pm2 的使用方法： [https://blog.csdn.net/KimBing/article/details/124249590](https://blog.csdn.net/KimBing/article/details/124249590)

```bash
pm2 start bin/www --name diary
```

如果你实在不用 pm2，也可以直接使用 `npm` 原始方法启动

```bash
npm run start
```

项目启动后会运行在 `localhost:3000`，直接访问这个地址应该能看到：

<img width="539" alt="Screen Shot 2022-04-19 at 21 47 25" src="https://user-images.githubusercontent.com/12215982/164018379-bb497ec3-53e4-46c5-969a-c5b3ca4c0c31.png">


### 4. 初始化数据库

直接访问 `你服务器的域名或IP:3000/init` 这个路径即可将数据库初始化。
初始化数据库会自动创建一个为名 `diary` 的数据库

初始化后，会自动在项目目录中新建一个名为 `DATABASE_LOCK` 的文件，之后将不能再执行这个接口，如果想要重新初始化，需要先将这个文件删除。

> 注意：初始化会清空 diary 数据库中的所有内容



### 5. 配置 nginx，映射 `localhost:3000` 路径到  `/portal` 路径

1. 打开 nginx 的配置文件，
- CentOS 的 nginx 配置文件在 `/etc/nginx/conf.d/` 目录下。
- Ubuntu 的 nginx 配置文件在 `/etc/nginx/site-avilable/default` 中。

2. 打开 `default.conf` 或 `default` 文件
    ```bash
    vi default.conf
    ```

3. 在 http 内部， server 外部，添加以下内容

    ```bash
     upstream diary_server {
         server localhost:3000;
         keepalive 2000;
     }
    ```

4. 然后在 server 内部添加：

    ```bash
    location /portal/ {
        proxy_pass http://diary_server/; # 这里对应上面的 upstream 名字
        proxy_set_header Host $host:$server_port; # 这里照搬就可以
    }
    ```
5. 这样，就会将 `localhost:3000` 这个接口映射到 `localhost/portal/` 这个路径下
6. 重启 nginx 服务
    ```bash
    systemctl restart nginx
    ```

### 6. 配置前端项目
1. 下载 [https://github.com/KyleBing/diary-vue](https://github.com/KyleBing/diary-vue)
2. 安装依赖，执行 `npm i` 或者 `yarn`
3. 如果你需要修改前端请求后台的路径，修改 `/src/request.js` 中的 `BASE_URL` 即可
4. 执行 `npm build` 生成最终项目文件
5. 放置于服务器 nginx 主目录下的 `/diary/` 下
6. 此时后台项目在 `/portal/` 目录下，前端项目在 `/dairy/` 下，这样就能直接使用了


## 三、开发说明

### 1. 密码说明
密码使用 [bcrypt](https://github.com/kelektiv/node.bcrypt.js) 加密，登录后的所有传输都会用加密后的密码作为请求 token

### 2. 返回数据格式

```json
{
  "success": true,
  "message": "提示信息",
  "data": {}
}
```

## 四、其它
> 始于 `2022-04-14`


CHANGELOG.md
# CHANGELOG



### v1.00 `2022-06-22`
- 添加账单数据，月份分类


routes/user/user.js
const express = require('express')
const configProject = require('../../config/configProject')
const utility = require("../../config/utility");
const ResponseSuccess = require("../../response/ResponseSuccess");
const ResponseError = require("../../response/ResponseError");
const router = express.Router()
const bcrypt = require('bcrypt')



/* GET users listing. */
router.post('/register', (req, res, next) => {
    // TODO: 验证传过来的数据库必填项
    // 判断邀请码是否正确
    if (req.body.invitationCode && req.body.invitationCode === configProject.invitation){
        checkEmailOrUserNameExist(req.body.email, req.body.username)
            .then(dataEmailExistArray => {
                // email 记录是否已经存在
                if (dataEmailExistArray.length > 0){
                    return res.send(new ResponseError('', '邮箱或用户名已被注册'))
                } else {
                    let sqlArray = []
                    let timeNow = utility.dateFormatter(new Date())
                    // 明文密码通过 bcrypt 加密，对比密码也是通过  bcrypt
                    bcrypt.hash(req.body.password, 10, (err, encryptPassword) => {
                        sqlArray.push(
                            // 注册的用户默认为普通用户
                            `insert into users(email, nickname, username, password, register_time, last_visit_time, comment, 
                                                wx, phone, homepage, gaode, group_id)
                                    VALUES (
                                    '${req.body.email}', 
                                    '${req.body.nickname || ''}', 
                                    '${req.body.username || ''}', 
                                    '${encryptPassword}', 
                                    '${timeNow}',
                                    '${timeNow}',
                                    '${req.body.comment || ''}', 
                                    '${req.body.wx || ''}', 
                                    '${req.body.phone || ''}', 
                                    '${req.body.homepage || ''}', 
                                    '${req.body.gaode || ''}', 
                                    '2'
                                    )`
                        )
                        utility
                            .getDataFromDB( 'diary', sqlArray)
                            .then(data => {
                                res.send(new ResponseSuccess('', '注册成功'))
                            })
                            .catch(err => {
                                res.send(new ResponseError('', '注册失败'))
                            })
                    })

                }
            })
            .catch(errEmailExist => {
                console.log(errEmailExist)
                res.send(new ResponseError(errEmailExist, '查询出错'))
            })

    } else {
        res.send(new ResponseError('', '邀请码错误'))
    }


})

// 检查用户名或邮箱是否存在
function checkEmailOrUserNameExist(email, username){
    let sqlArray = []
    sqlArray.push(`select * from users where email='${email}' or username ='${username}'`)
    return utility.getDataFromDB( 'diary', sqlArray)
}

router.get('/list', (req, res, next) => {
    utility
        .verifyAuthorization(req)
        .then(userInfo => {
            let sqlArray = []
            sqlArray.push(`SELECT * from users `)

            if (userInfo.group_id === 1){

            } else {
                sqlArray.push([`where uid = ${userInfo.uid}`])
            }


            // keywords
            if (req.query.keywords){
                let keywords = JSON.parse(req.query.keywords).map(item => utility.unicodeEncode(item))
                console.log(keywords)
                if (keywords.length > 0){
                    let keywordStrArray = keywords.map(keyword => `( message like '%${keyword}%' ESCAPE '/'  or description like '%${keyword}%' ESCAPE '/')` )
                    sqlArray.push(' and ' + keywordStrArray.join(' and ')) // 在每个 categoryString 中间添加 'or'
                }
            }

            if (req.query.pageNo && req.query.pageSize){
                let startPoint = (req.query.pageNo - 1) * req.query.pageSize //  QR 起点
                sqlArray.push(`limit ${startPoint}, ${req.query.pageSize}`)
            }

            utility
                .getDataFromDB( 'diary', sqlArray)
                .then(dataDiary => {
                    utility.updateUserLastLoginTime(userInfo.uid)
                    dataDiary.forEach(diary => {
                        // decode unicode
                        diary.title = utility.unicodeDecode(diary.title)
                        diary.content = utility.unicodeDecode(diary.content)
                    })
                    res.send(new ResponseSuccess(dataDiary, '请求成功'))
                })
                .catch(err => {
                    res.send(new ResponseError(err, err.message))
                })
        })
        .catch(verified => {
            res.send(new ResponseError(verified, '无权查看用户列表：用户信息错误'))
        })
})

router.get('/detail', (req, res, next) => {
    let sqlArray = []
    sqlArray.push(`select * from qrs where hash = '${req.query.hash}'`)
    utility
        .getDataFromDB( 'diary', sqlArray, true)
        .then(data => {
            // decode unicode
            data.message = utility.unicodeDecode(data.message)
            data.description = utility.unicodeDecode(data.description)

            // 2. 判断是否为共享 QR
            if (data.is_public === 1){
                // 2.1 如果是，直接返回结果，不需要判断任何东西
                res.send(new ResponseSuccess(data))
            } else {
                // 2.2 如果不是，需要判断：当前 email 和 token 是否吻合
                utility
                    .verifyAuthorization(req)
                    .then(userInfo => {
                        // 3. 判断 QR 是否属于当前请求用户
                        if (Number(userInfo.uid) === data.uid){
                            // 记录最后访问时间
                            utility.updateUserLastLoginTime(userInfo.uid)
                            /*                            // TODO:过滤可见信息 自己看，管理员看，其它用户看
                                                        if (data.is_show_wx){
                                                            data.wx = ''
                                                        }
                                                        if (data.is_show_car){
                                                            data.car = ''
                                                            data.car_desc = ''
                                                            data.car_plate = ''
                                                        }
                                                        if (data.is_show_gaode){
                                                            data.gaode = ''
                                                        }
                                                        if (data.is_show_homepage){
                                                            data.homepage = ''
                                                        }*/
                            res.send(new ResponseSuccess(data))
                        } else {
                            res.send(new ResponseError('','当前用户无权查看该 QR ：请求用户 ID 与 QR 归属不匹配'))
                        }
                    })
                    .catch(unverified => {
                        res.send(new ResponseError('','当前用户无权查看该 QR ：用户信息错误'))
                    })
            }
        })
        .catch(err => {
            res.send(new ResponseError(err, err.message))
        })
})

router.post('/add', (req, res, next) => {
    checkEmailOrUserNameExist(req.body.email, req.body.username)
        .then(dataEmailExistArray => {
            // email 记录是否已经存在
            if (dataEmailExistArray.length > 0){
                return res.send(new ResponseError('', '邮箱或用户名已被注册'))
            } else {
                let sqlArray = []
                let timeNow = utility.dateFormatter(new Date())
                // 明文密码通过 bcrypt 加密，对比密码也是通过  bcrypt
                bcrypt.hash(req.body.password, 10, (err, encryptPassword) => {
                    sqlArray.push(
                        `insert into users(email, nickname, username, password, register_time, last_visit_time, comment, 
                                                wx, phone, homepage, gaode, group_id)
                                    VALUES (
                                    '${req.body.email}', 
                                    '${req.body.nickname}', 
                                    '${req.body.username}', 
                                    '${encryptPassword}', 
                                    '${timeNow}',
                                    '${timeNow}',
                                    '${req.body.comment || ''}', 
                                    '${req.body.wx}', 
                                    '${req.body.phone}', 
                                    '${req.body.homepage}', 
                                    '${req.body.gaode}', 
                                    '${req.body.group_id}'
                                    )`
                    )
                    utility
                        .getDataFromDB( 'diary', sqlArray)
                        .then(data => {
                            res.send(new ResponseSuccess('', '用户添加成功'))
                        })
                        .catch(err => {
                            res.send(new ResponseError(err, '用户添加失败'))
                        })
                })

            }
        })
        .catch(errEmailExist => {
            console.log(errEmailExist)
            res.send(new ResponseError(errEmailExist, '查询出错'))
        })
})

// 检查用户名或邮箱是否存在
function checkHashExist(username, email){
    let sqlArray = []
    sqlArray.push(`select * from users where username ='${username.toLowerCase()}' or email = '${email}'`)
    return utility.getDataFromDB( 'diary', sqlArray)
}

router.put('/modify', (req, res, next) => {

    // 1. 验证用户信息是否正确
    utility
        .verifyAuthorization(req)
        .then(userInfo => {
            if (userInfo.group_id === 1) {
                operateUserInfo(req, res, userInfo)
            } else {
                if (userInfo.uid !== req.body.uid){
                    res.send(new ResponseError('', '你无权操作该用户信息'))
                } else {
                    operateUserInfo(req, res)
                }
            }
        })
        .catch(err => {
            res.send(new ResponseError(err, '无权操作'))
        })
})

function operateUserInfo(req, res, userInfo){
    let sqlArray = []
    sqlArray.push(`
                        update users
                            set
                                    users.email = '${req.body.email}', 
                                    users.nickname = '${req.body.nickname}', 
                                    users.username = '${req.body.username}', 
                                    users.comment = '${req.body.comment || ''}', 
                                    users.wx = '${req.body.wx}', 
                                    users.phone = '${req.body.phone}', 
                                    users.homepage = '${req.body.homepage}', 
                                    users.gaode = '${req.body.gaode}', 
                                    users.group_id = '${req.body.group_id}'
                            WHERE uid='${req.body.uid}'
                    `)

    utility
        .getDataFromDB( 'diary', sqlArray, true)
        .then(data => {
            utility.updateUserLastLoginTime(userInfo.uid)
            res.send(new ResponseSuccess(data, '修改成功'))
        })
        .catch(err => {
            res.send(new ResponseError(err, '修改失败'))
        })
}

router.delete('/delete', (req, res, next) => {
    // 1. 验证用户信息是否正确
    utility
        .verifyAuthorization(req)
        .then(userInfo => {
            if (userInfo.group_id === 1){
                let sqlArray = []
                sqlArray.push(`
                        DELETE from users
                        WHERE uid='${req.body.uid}'
                    `)
                utility
                    .getDataFromDB( 'diary', sqlArray)
                    .then(data => {
                        if (data.affectedRows > 0) {
                            utility.updateUserLastLoginTime(req.body.email)
                            res.send(new ResponseSuccess('', '删除成功'))
                        } else {
                            res.send(new ResponseError('', '删除失败'))
                        }
                    })
                    .catch(err => {
                        res.send(new ResponseError(err,))
                    })
            } else {
                res.send(new ResponseError('', '无权操作'))
            }
        })
        .catch(err => {
            res.send(new ResponseError(err, '无权操作'))
        })
})

router.post('/login', (req, res, next) => {
    let sqlArray = []
    sqlArray.push(`select * from users where email = '${req.body.email}'`)

    utility
        .getDataFromDB( 'diary', sqlArray, true)
        .then(data => {
            if (data) {
                bcrypt.compare(req.body.password, data.password, function(err, isPasswordMatch) {
                    if (isPasswordMatch){
                        utility.updateUserLastLoginTime(req.body.email)
                        res.send(new ResponseSuccess(data,'登录成功'))
                    } else {
                        res.send(new ResponseError('','用户名或密码错误'))
                    }
                })
            } else {
                res.send(new ResponseError('', '无此用户'))
            }

        })
        .catch(err => {
            res.send(new ResponseError('', err.message))
        })
})

// 修改密码
router.put('/change-password', (req, res, next) => {
    if (!req.body.password){
        res.send(new ResponseError('', '参数错误：password 未定义'))
        return
    }
    if (req.query.email === 'test@163.com'){
        res.send(new ResponseError('', '演示账户密码不允许修改'))
        return
    }
    utility
        .verifyAuthorization(req)
        .then(userInfo => {
            if (userInfo.password === req.query.token){
                bcrypt.hash(req.body.password, 10, (err, encryptPasswordNew) => {
                    let changePasswordSqlArray = [`update users set password = '${encryptPasswordNew}' where email='${req.query.email}'`]
                    utility
                        .getDataFromDB( 'diary', changePasswordSqlArray)
                        .then(dataChangePassword => {
                            utility.updateUserLastLoginTime(userInfo.uid)
                            res.send(new ResponseSuccess('', '修改密码成功'))
                        })
                        .catch(errChangePassword => {
                            res.send(new ResponseError('', '修改密码失败'))
                        })
                })
            }
        })
        .catch(err => {
            res.send(new ResponseError(err, '无权操作'))
        })

})


module.exports = router
